### PR TITLE
Add filament usage stats to generated GCode files

### LINKED ISSUE #3072: Missing properties from sliced files
**Is your feature request related to a problem? Please describe.**
I'm building a business using Bambu Printers, and I want to automatize the slicing process. However there are a few properties missing from the sliced gcode output which was available in the upstream repository of Bambu Studio and are required for me to properly calculate the prices of prints.

**Describe the solution you'd like**
I would like to get back the following proprties in the GCode output (header comments):
- filament used [mm] 
- filament used [cm3]

**Describe alternatives you've considered**
For me, I don't see any alternatives. I'm forced to use Prusa Slicer (the upstream repository) until this gets resolved, which means I need to convert config files, etc. etc.

### PR DESCRIPTION
**Pull request by THE-SIMPLE-MARK**
This feature is available in Slic3r and Prusa slicer, however it seems that this functionality has never been migrated to Bambu Studio. This PR aims to do exactly that.

This would fix #3072.
This would close #2879.

### To-do list
- [x] Add placeholders for properties
- [x] Fill in placeholders with actual data in post processing (in `process_gcode_line`)
  - [x] Total filament used in **grams**
  - [x] Total filament volume used in $cm^3$
  - [x] Total filament length used in **millimeters**

### CODE REVIEW

**Comment by THE-SIMPLE-MARK** [comment]
@XunZhangBambu Here you go.

I'm going to remove the placeholder for the unnecessary filament cost property (unable to be set up from CLI). But that's it from my part for now.

**Comment by THE-SIMPLE-MARK** [comment]
I think a review at this stage is unnecessary because there are still quite many things to do.

**Comment by XunZhangBambu** [comment]
@THE-SIMPLE-MARK  Hi, I have pushed the latest commit to branch "filament"

**Comment by THE-SIMPLE-MARK** [comment]
> @THE-SIMPLE-MARK Hi, I have pushed the latest commit to branch "filament"

Awesome! May I ask what did you add? Might be able to tick off a few of our tasks.

**Comment by XunZhangBambu** [comment]
@THE-SIMPLE-MARK  Of course. Add the data you mentioned above.  
Total filament used in grams
Total filament volume used in grams
Total filament length used in milimeters

**Comment by THE-SIMPLE-MARK** [comment]
> @THE-SIMPLE-MARK Of course. Add the data you mentioned above. Total filament used in grams Total filament volume used in grams Total filament length used in milimeters

@XunZhangBambu Alright, then I think we are ready for review!

**Comment by THE-SIMPLE-MARK** [comment]
@SaltWei @XunZhangBambu Could we get this reviewed please? I'm really looking to see this in the next release of Bambu Studio as it has been requested by many people.

**Comment by SaltWei** [comment]
> @SaltWei @XunZhangBambu Could we get this reviewed & merged please? I'm really looking to see this in the next release of Bambu Studio as it has been requested by many people.

Yes. Will be finished soon

File: src/libslic3r/GCode/GCodeProcessor.cpp [hunk 1]
```diff
@@ -59,8 +59,14 @@ const std::vector<std::string> GCodeProcessor::Reserved_Tags = {
     "_GP_LAST_LINE_M73_PLACEHOLDER",
     "_GP_ESTIMATED_PRINTING_TIME_PLACEHOLDER",
     "_GP_TOTAL_LAYER_NUMBER_PLACEHOLDER",
+    "_GP_TOTAL_FILAMENT_USED_MM_PLACEHOLDER",
```
Comment (XunZhangBambu) [inline src/libslic3r/GCode/GCodeProcessor.cpp:62]: Used tags

File: src/libslic3r/GCode/GCodeProcessor.cpp [hunk 2]
```diff
@@ -59,8 +59,14 @@ const std::vector<std::string> GCodeProcessor::Reserved_Tags = {
...
+    "_GP_TOTAL_FILAMENT_USED_CM3_PLACEHOLDER",
```
Comment (XunZhangBambu) [inline src/libslic3r/GCode/GCodeProcessor.cpp:63]: The same

File: src/libslic3r/GCode/GCodeProcessor.cpp [hunk 3]
```diff
@@ -59,8 +59,14 @@ const std::vector<std::string> GCodeProcessor::Reserved_Tags = {
...
+    "_GP_TOTAL_FILAMENT_USED_G_PLACEHOLDER",
```
Comment (XunZhangBambu) [inline src/libslic3r/GCode/GCodeProcessor.cpp:64]: Same

**Review by XunZhangBambu — CHANGES_REQUESTED** [review]

File: src/libslic3r/GCode/GCodeProcessor.cpp (hunk 3 above)
Comment (THE-SIMPLE-MARK) [reply src/libslic3r/GCode/GCodeProcessor.cpp:64]: > Same

@XunZhangBambu Would you like me to remove all of these entries from 59-64?

**Comment by THE-SIMPLE-MARK** [comment]
@XunZhangBambu I have implemented the requested changes.

**Review by XunZhangBambu — APPROVED** [review]

**Comment by THE-SIMPLE-MARK** [comment]
@XunZhangBambu Do we know when this will be merged? Would love to see this in the main release. Or has it been already implemented internally?

**Comment by epitaphmike** [comment]
Any update on when this might make it into the main release @SaltWei? Thank you everyone for your time spent on this!

**Comment by TheOddDuck** [comment]
@XunZhangBambu @SaltWei @THE-SIMPLE-MARK 

Any updates on this? I am also looking for this functionality and it looks like this has been pending a second code review for several months.

**Comment by XunZhangBambu** [comment]
@THE-SIMPLE-MARK  We are about to merge this PR, but since it has been a long time, some modifications need to be made to the code. Can you synchronize with our latest code? After synchronization, I will submit some code modifications to your PR.

**Comment by THE-SIMPLE-MARK** [comment]
> @THE-SIMPLE-MARK We are about to merge this PR, but since it has been a long time, some modifications need to be made to the code. Can you synchronize with our latest code? After synchronization, I will submit some code modifications to your PR.

@XunZhangBambu Done.

**Comment by XunZhangBambu** [comment]
@THE-SIMPLE-MARK  Can you create a new branch based on the latest version of bambu studio? I will integrate all the changes into one commit and submit it to the new branch. I prefer this approach more. Thank you!

**Comment by THE-SIMPLE-MARK** [comment]
> @THE-SIMPLE-MARK Can you create a new branch based on the latest version of bambu studio? I will integrate all the changes into one commit and submit it to the new branch. I prefer this approach more. Thank you!

@XunZhangBambu Done. The new branch is available under [THE-SIMPLE-MARK:filament-usage-stats](https://github.com/THE-SIMPLE-MARK/BambuStudio/tree/filament-usage-stats).

**Comment by TheOddDuck** [comment]
@XunZhangBambu - This is fantastic! Any idea on when this will be fully approved and in a potential release? REALLY looking forward to this!

**Comment by XunZhangBambu** [comment]
@THE-SIMPLE-MARK  Already pushed a new commit to branch "filament-usage-stats"!

**Comment by XunZhangBambu** [comment]
@THE-SIMPLE-MARK   Do you mind if we submit the contents of this PR to our internal code repository and change the author to your name?

**Comment by THE-SIMPLE-MARK** [comment]
> @THE-SIMPLE-MARK Do you mind if we submit the contents of this PR to our internal code repository and change the author to your name?

Sure. Whatever it takes to get this merged.

**Comment by XunZhangBambu** [comment]
@THE-SIMPLE-MARK  @epitaphmike  This feature has been released in the latest version. I think we can close this pull request now.

**Comment by THE-SIMPLE-MARK** [comment]
Alright

**Pull request closed without merging**